### DISCUSSION TITLE
Django + Gunicorn: How to go about running shared theading.Thread() + subprocess.popen() on startup

### DISCUSSION THREAD

**Discussion by cerahmed** [benoitc/gunicorn#3202] [Q&A] [answered]
Hi All,

I have a django backend that, at startup, does two things:

1) Runs a `threading.Thread()` that calls a function every 10 seconds to read RPi's sensor information and saves it to database.
2) Runs a `subprocess.Popen()` to call a binary linux program (**go2rtc** to be specific) that starts the streaming service indefinetely until it's manually stopped.

These calls are performed in each django app's AppConfig `def ready()` method.

When running django's development server, everything works as expected. However, when running through Gunicorn with multiple workers, things get a bit messy.

I'd like some help in the following:

1) With multiple workers, the `def ready()` method gets called for each worker, resulting in redundant creation of threads and Popen processes. Using `--preload` seems to solve the issue. However, the `threading.Thread()` output to journalctl seems to be blocked; only shows in journalctl after gunicorn is stopped/restared. What is going on here?

2) I need the Thread and Popen instances to be accessible by all workers so that when I send a `stop` request from the frontend, these threads/processes can halt gracefully. How would I go about sharing these instances between all the workers?

My current gunicorn service is as follows:

```
[Service]
User=pi
Group=www-data
WorkingDirectory=/home/pi/homecontrol
ExecStart=/home/pi/homecontrol/venv/bin/gunicorn \
          --access-logfile - \
          --workers 3 \
          --bind unix:/run/gunicorn.sock \
          --preload \
          homecontrol.wsgi:application
```

Appreciate some insights on this.

**Comment by pajod** [comment]
Sounds like there is nothing in the threads dealing with external sensors/executables that require it to share process/thread context with the HTTP serving, other than managing start/stop of dependent services. Which is one of the key features of systemd, which you already use.

Why have it interfere with your flexibility in controlling HTTP thread/process management, then? Wrap it inside what Django calls "management commands" and have those launched by entirely separate systemd services, declaring dependencies as needed to have systemd manage launching / stopping it.

**Reply by cerahmed** [reply]
Hey @pajod, thanks for taking some time to help.

Let me try to clarify my use case for the go2rtc instance, which will help me debug both issues:

1) At startup, I spawn a new process with `subprocess.Popen` to run `go2rtc` binary (which is blocking by default). This program start a streaming service at port 1984 and listens for peer connections indefinetely.

2) In my frontend, the admin can trigger a start, stop, restart go2rtc service through a custom API endpoint. For these actions to take action, they need a reference to the subprocess.Popen instance that was created at startup, in order to SIGTERM it. 

3) Currently, I have the Popen instance in a global variable (not thread safe). With many workers, only one worker has the Popen instance and can signal termination, for other workers Popen instance is None.

4) Although I'm aware of management commands, I'm not sure how I can access this Popen instance from all workers. If you have reading resources that would be appreciated.

**Reply by pajod** [reply]
That still sounds like implementing service management scaffolding in Python, when you already use a dedicated service management system that can be configured for most common requirements.

You are not exchanging further information that would require direct access to the Popen object, do you? So you want to SIGTERM an external dependency of your service? You can typically signal *any* process of the same namespace+user - even a service you only learned the process ID of, after it was started by the system manager in your name (try it: `$ kill -TERM $(systemctl show --property=MainPID --value my-external-service.service)`). Systemd can start/reload the external binary directly e.g. by watching completed writes to that programs configuration file (writing which would be then what is left to do from the gunicorn side). It will also know to distinguish an error exit code (and arrange automatic restart, where appropriate) and a successful exit (after receiving SIGTERM from any of your gunicorn-hosted threads).

This is what I imagine your IPC channels look like, after:
* HTTP server -> writes external-binary.config -> systemd watching that file starts/restarts `/path/to/external-binary`
* HTTP server -> sends SIGTERM for external-binary -> `/path/to/external-binary` stops
* HTTP server -> reads db
* sensor polling service -> writes to db
* `/path/to/external-binary` occasionally processes signals, knows little and cares little about how exactly it is being controlled

Assuming that is a fair assessment, I see nothing in requiring any tighter integration than gunicorn running in one systemd service (but possibly many threads and processes) and a few external services running in separate systemd services.

**Reply by cerahmed** [reply]
Awsome breakdown for the overall architecture of the issue; really apprecaite it.

One point I'm missing which you slightly touched upon is where (considering best practices) should I store the PID of the external process? Should it be in the DB so that all Gunicorn workers can get it?

**Reply by pajod** [reply]
> should I store the PID

I don't, but have not spent much thought on that. In my case, not a performance issue to query (by name, or to systemd) every time.
Note that retrieving the PID and then separately signalling (potentially since-recycled) that PID is a race condition. If you want to store it, minimize that window in the service crash case by storing the PID in a text file, and instructing systemd to *delete* that file once the service exits (using `PIDFile=`).

**Reply by cerahmed** [reply]
Querying the PID by service name seems to be the best way to go.

If you don't mind helping with the other issue now that the subprocess issue is out of the question.

At startup of my django project, I instantiate a [pyhap ](https://github.com/ikalchev/HAP-python) driver object, which creates homekit accessories and broadcast the service to the network.

Unlike the subprocess binary program, this object must stay within django in order to dynamically exchange information with the front end. Again, in my MVP django project, the driver instance lives in the program as a global variable, which the frontend can access through API endpoints to start/stop/update using the driver's instance methods. Also, the driver object needs to continuously make calls to the django's DB in order to make changes as necessary.

What is the best practice for sharing the driver object between workers? I've read about Celery and Redis, but not sure if that's the easiest way to go about it.

**Reply by pajod** [reply, marked as answer]
An extra web service that you do want to reconfigure on the python side? That one sounds more like an application for using a threaded worker model in gunicorn, and setting up / tearing down the shared object from the gunicorn *server hooks*

**Reply by cerahmed** [reply]
gthreads might actually be the most suitable answer to all my concerns!

I went ahead and read about the different worker types. Ended up specifying only 1 worker, 8 threads, and removed the preload directive; this caused everything to work as expected. I'm quite new to gunicorn so I'm not sure about the side effects this have over running single-threaded multiple workers, but this will do for now.

`gunicorn --bind 0.0.0.0:8000 --workers 1 --thread 8 --access-logfile - homecontrol.wsgi:application --graceful-timeout 5`

I'm sure that it'd be wiser to adjust my django project for separate django commands for scalability, but that's a problem for another day.

Thanks a lot @pajod for the great support; very much appreciated.
